### PR TITLE
added sparkmax smart motion pid, cr needed

### PR DESCRIPTION
**Pull request by benalaluf**

### CODE REVIEW

File: src/pid/RevController.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        sparkMax.getPIDController().setD(pidfTerms.getKd());
+        sparkMax.getPIDController().setFF(pidfTerms.getKf());
+    }
+
+    protected boolean isFirstRun() {
+        return getCurrentError() == firstError;
+    }
+
+    protected void setReference(double setpoint, CANSparkMax.ControlType controlType, int slotId, double feedForward) {
+        this.sparkMax.getPIDController().setReference(setpoint / 42, controlType, slotId, feedForward);
```
Comment (Romeore) [inline src/pid/RevController.java:63]: Make a constant for this magic number

File: src/pid/RevController.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    protected void setReference(double setpoint, CANSparkMax.ControlType controlType) {
+        setReference(setpoint,controlType,0,0);
```
[the remark this answers is not in the document]
Comment (Romeore) [reply src/pid/RevController.java:67]: ctrl alt l

File: src/pid/RevSmartMotionController.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    public RevSmartMotionController(CANSparkMax sparkMax, double kP,
+                                    double kI, double kD, double kF, int slotIdx,
+                                    int maxAcceleration, int maxVelocity, int minVelocity) {
+        super(sparkMax, kP, kI, kD, kF, slotIdx);
+        this.sparkMaxPIDController = sparkMax.getPIDController();
+        setMaxAcceleration(maxAcceleration);
+        setMaxAndMinVelocity(maxVelocity, minVelocity);
+        setOutPutRange(-1, 1);
```
Comment (Romeore) [inline src/pid/RevSmartMotionController.java:40]: What is `-1,1`

**Review by Romeore — CHANGES_REQUESTED** [review]

File: src/pid/RevController.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    protected boolean isFirstRun() {
+        return getCurrentError() == firstError;
+    }
+
+    protected void setReference(double setpoint, CANSparkMax.ControlType controlType, int slotId, double feedForward) {
+        this.sparkMax.getPIDController().setReference(setpoint / DEFAULT_ENCODER_UNITS_PER_ROTATION, controlType, slotId, feedForward);
+    }
+
+    protected void setReference(double setpoint, CANSparkMax.ControlType controlType) {
+        setReference(setpoint, controlType, ZERO, ZERO);
```
Comment (nimrod46) [inline src/pid/RevController.java:66]: this should be `setReference(setpoint, controlType, slotId, 0);`

**Review by nimrod46 — CHANGES_REQUESTED** [review]

File: src/pid/RevController.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        sparkMax.getPIDController().setFF(pidfTerms.getKf());
+    }
+
+    protected boolean isFirstRun() {
+        return getCurrentError() == firstError;
+    }
+
+    protected void setReference(double setpoint, CANSparkMax.ControlType controlType, int slotId, double feedForward) {
+        this.sparkMax.getPIDController().setReference(setpoint / DEFAULT_ENCODER_UNITS_PER_ROTATION, controlType, slotId, feedForward);
+    }
```
Comment (nimrod46) [inline src/pid/RevController.java:63]: Remove `slotId` as a argument, use the field this object has instead

File: src/pid/RevSmartMotionController.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    public RevSmartMotionController(CANSparkMax sparkMax,
+                                    PIDFTerms pidfTerms, int slotIdx, int maxAcceleration,
+                                    int maxVelocity, int minVelocity) {
+        this(sparkMax, pidfTerms.getKp(), pidfTerms.getKi(), pidfTerms.getKd(),
+                pidfTerms.getKf(), slotIdx, maxAcceleration, maxVelocity, minVelocity);
+    }
+
+    public RevSmartMotionController(CANSparkMax sparkMax, double kP,
+                                    double kI, double kD, double kF, int slotIdx,
+                                    int maxAcceleration, int maxVelocity, int minVelocity) {
```
Comment (nimrod46) [inline src/pid/RevSmartMotionController.java:41]: int or double?

File: src/pid/RevSmartMotionController.java [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        this.sparkMaxPIDController = sparkMax.getPIDController();
+        this.maxAcceleration = maxAcceleration;
+        this.maxVelocity = maxVelocity;
+        this.minVelocity = minVelocity;
+        this.minOutput = sparkMaxPIDController.getOutputMin();
+        this.maxOutput = sparkMaxPIDController.getOutputMax();
+        this.kIZ = sparkMaxPIDController.getIZone();
+        setMaxAcceleration(maxAcceleration);
+        setMaxAndMinVelocity(maxVelocity, minVelocity);
+        setOutputRange(DEFAULT_MIN_OUTPUT, DEFAULT_MAX_OUTPUT);
```
Comment (nimrod46) [inline src/pid/RevSmartMotionController.java:46]: Dot set it to an arbitrary value, use the value already inside `sparkMaxPIDController `

File: src/pid/RevConstant.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+    public static final double DEFAULT_MIN_OUTPUT = -1;
+
+    public static final double DEFAULT_MAX_OUTPUT = 1;
+
+    public static final int DEFAULT_SLOT_ID = 0;
+
+    public static final int DEFAULT_ENCODER_UNITS_PER_ROTATION = 42;
+
+    public static final int ZERO = 0;
```
Comment (nimrod46) [inline src/pid/RevConstant.java:13]: Remove

**Review by nimrod46 — CHANGES_REQUESTED** [review]

File: src/pid/RevSmartMotionController.java [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+    public void setMaxAndMinVelocity(double maxVelocity, double minVelocity) {
+        this.maxVelocity = maxVelocity;
+        this.maxVelocity = minVelocity;
+        this.sparkMaxPIDController.setSmartMotionMaxVelocity(maxVelocity, slotId);
+        this.sparkMaxPIDController.setSmartMotionMinOutputVelocity(minVelocity, slotId);
+    }
+
+    public void setTolerance(double tolerance) {
+        this.sparkMaxPIDController.setSmartMotionAllowedClosedLoopError(tolerance, slotId);
```
Comment (nimrod46) [inline src/pid/RevSmartMotionController.java:134]: Create `tolerance` field and update it here as well

**Review by nimrod46 — CHANGES_REQUESTED** [review]

File: src/pid/RevConstant.java [hunk 10]
```diff
@@ -0,0 +1,13 @@
+package pid;
+
+public class RevConstant {
```
Comment (Romeore) [inline src/pid/RevConstant.java:3]: `final` class, rename to `RevConstants`

File: src/pid/RevController.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+        this.setPIDFTerms(pidfTerms);
+    }
+
+    public abstract void update(double setpoint, double feedForward);
+
+    public void resetEncoder() {
+        revEncoder.reset();
+    }
+}
+//by.Ben
```
Comment (Romeore) [inline src/pid/RevController.java:93]: remove

File: src/pid/RevSmartMotionController.java [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+public class RevSmartMotionController extends RevController {
+
+    private final SparkMaxPIDController controller;
+    private double maxAcceleration;
+    private double maxVelocity;
+    private double minVelocity;
+    private double minOutput;
+    private double maxOutput;
+    private double kIZ;
+    private int tolerance;
```
Comment (Romeore) [inline src/pid/RevSmartMotionController.java:17]: This is not the order according to conventions

File: src/pid/RevController.java [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+import static pid.RevConstant.DEFAULT_SLOT_ID;
+
+public abstract class RevController extends AbstractController {
+
+    protected CANSparkMax sparkMax;
+    protected RevEncoder revEncoder;
+    protected int slotId;
+    protected double firstError;
+
+    public RevController(CANSparkMax canSparkMax,
```
Comment (Romeore) [inline src/pid/RevController.java:16]: Why do we call this parameter `canSparkMax` on 2 constructors and `sparkMax` on the other ones?

File: src/pid/RevSmartMotionController.java [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+
+    public double getMinOutputRange() {
+        return controller.getOutputMin();
+    }
+
+    public double getMaxOutputRange() {
+        return controller.getOutputMax();
+    }
+}
+//by.Ben
```
Comment (Romeore) [inline src/pid/RevSmartMotionController.java:157]: Remove..

File: src/pid/RevSmartMotionController.java [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    @Override
+    protected void configVariables() {
+        super.configVariables();
+        setMaxAcceleration(maxAcceleration);
+        setMaxAndMinVelocity(maxVelocity, minVelocity);
+        setOutputRange(minOutput, maxOutput);
+        setIZone(kIZ);
+
```
Comment (Romeore) [inline src/pid/RevSmartMotionController.java:111]: Remove space

File: src/pid/RevSmartMotionController.java [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+                pidfTerms.getKf(), slotId, maxAcceleration, maxVelocity, minVelocity, tolerance);
+    }
+
+    public RevSmartMotionController(CANSparkMax sparkMax, double kP,
+                                    double kI, double kD, double kF, int slotId,
+                                    double maxAcceleration, double maxVelocity, double minVelocity, int tolerance) {
+        super(sparkMax, kP, kI, kD, kF, slotId);
+        this.controller = sparkMax.getPIDController();
+        setMaxAcceleration(maxAcceleration);
+        setMaxAndMinVelocity(maxVelocity,minVelocity);
```
Comment (Romeore) [inline src/pid/RevSmartMotionController.java:45]: Reformat

File: src/pid/RevSmartMotionController.java [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+    public RevSmartMotionController(CANSparkMax sparkMax,
+                                    PIDFTerms pidfTerms, int slotId, double maxAcceleration,
+                                    double maxVelocity, double minVelocity, int tolerance) {
+        this(sparkMax, pidfTerms.getKp(), pidfTerms.getKi(), pidfTerms.getKd(),
+                pidfTerms.getKf(), slotId, maxAcceleration, maxVelocity, minVelocity, tolerance);
+    }
+
+    public RevSmartMotionController(CANSparkMax sparkMax, double kP,
+                                    double kI, double kD, double kF, int slotId,
+                                    double maxAcceleration, double maxVelocity, double minVelocity, int tolerance) {
```
Comment (Romeore) [inline src/pid/RevSmartMotionController.java:41]: Get the minVelocity first and then the maxVelocity
Fix at all constructors.

**Review by Romeore — CHANGES_REQUESTED** [review]

**Review by Romeore — APPROVED** [review]

File: src/sensors/counter/RevAlternateEncoder.java [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+
+    @Override
+    public double getRate() {
+        return sparkMax.getAlternateEncoder(revCounterType.getEncoderUnitsPerRotation()).getVelocity();
+    }
+
+    @Override
+    public void reset() {
+        sparkMax.getAlternateEncoder(revCounterType.getEncoderUnitsPerRotation()).setPosition(0);
+    }
```
Comment (nimrod46) [inline src/sensors/counter/RevAlternateEncoder.java:27]: Do not call `getAlternateEncoder` multiple times, save the encoder instance in a field

File: src/pid/RevSmartMotionController.java [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    @Override
+    protected void configVariables() {
+        super.configVariables();
+        setMaxAcceleration(maxAcceleration);
+        setMinAndMaxVelocity(minVelocity, maxVelocity);
+        setOutputRange(minOutput, maxOutput);
+        setIZone(kIZ);
+        setTolerance(tolerance);
```
Comment (nimrod46) [inline src/pid/RevSmartMotionController.java:110]: Call the private function i mentioned earlier

File: src/pid/RevSmartMotionController.java [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+    public RevSmartMotionController(CANSparkMax sparkMax, RevCounter counter,
+                                    double kP, double kI, double kD, double kF, int slotId,
+                                    double maxAcceleration, double minVelocity, double maxVelocity, double tolerance) {
+        super(sparkMax, counter, kP, kI, kD, kF, slotId);
+        this.controller = sparkMax.getPIDController();
+        setMaxAcceleration(maxAcceleration);
+        setMinAndMaxVelocity(minVelocity, maxVelocity);
+        setOutputRange(controller.getOutputMin(), controller.getOutputMax());
+        setIZone(kIZ);
+        setTolerance(tolerance);
```
Comment (nimrod46) [inline src/pid/RevSmartMotionController.java:47]: Move to private function which will accept all of those arguments

**Review by nimrod46 — CHANGES_REQUESTED** [review]

**Review by nimrod46 — APPROVED** [review]

**Pull request merged**